### PR TITLE
Add cpu() and cuda() implementations to ATen

### PR DESCRIPTION
**Pull request by ezyang**
- to() is now correctly marked as this-const
- As native functions, cpu/cuda get automatically bound.  I didn't
  see a convenient way to maintain the 'async' kwarg, so it got
  dropped for now. This probably has to be fixed before we can merge this.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

CC @goldsborough

### CODE REVIEW

**Comment by SsnL** [comment]
We should make python `.to` (currently called via [here](https://github.com/pytorch/pytorch/blob/master/tools/autograd/templates/python_variable_methods.cpp#L568-L590)) and ATen `.to` go through the same code path one day.

File: aten/src/ATen/native/TensorFactories.cpp [hunk 1]
```diff
@@ -607,7 +623,7 @@ Tensor tensor_cpu(ArrayRef<T> values, const TensorOptions& options) {
 
 template <typename T>
 Tensor tensor_cuda(ArrayRef<T> values, const TensorOptions& options) {
-  auto cpu_tensor = tensor_cpu(values, TensorOptions(options).device(at::kCPU));
+  auto cpu_tensor = tensor_cpu(values, TensorOptions(options).device(Device(kCPU)));
```
Comment (goldsborough) [inline aten/src/ATen/native/TensorFactories.cpp]: Any reason for this?

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -1784,3 +1784,10 @@
 
 - func: get_device(Tensor self) -> int64_t
   device_guard: False
+
+- func: cpu(Tensor self) -> Tensor
+  variants: method
+
+# TODO: Would be preferable if this were an at::optional<int64_t> device
```
Comment (goldsborough) [inline aten/src/ATen/native/native_functions.yaml]: We need to fix optional on Windows first (believe me, I tried really hard and had to change it all back to -1 in the end...)

File: aten/src/ATen/templates/Tensor.h [hunk 3]
```diff
@@ -87,9 +87,10 @@ struct Tensor : public detail::TensorBase {
   inline Tensor toBackend(Backend b) const;
 
   /// New-style `to()` methods.
-  Tensor to(Device device, ScalarType dtype, bool non_blocking = false);
-  Tensor to(ScalarType dtype, bool non_blocking = false);
-  Tensor to(Device device, bool non_blocking = false);
+  /// NB: these are defined in TensorOptions.h
+  Tensor to(Device device, ScalarType dtype, bool non_blocking = false) const;
```
Comment (goldsborough) [inline aten/src/ATen/templates/Tensor.h]: Thanks!

File: tools/autograd/templates/python_variable_methods.cpp [hunk 4]
```diff
@@ -309,40 +309,6 @@ static PyObject * THPVariable_invert(PyObject* self, PyObject* args) {
...
-  auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
-  ParsedArgs<2> parsed_args;
-  auto r = parser.parse(args, kwargs, parsed_args);
-  auto backend = self_.is_sparse() ? at::kSparseCUDA : at::kCUDA;
-  auto& type = self_.type().toBackend(backend);
-  auto device_obj = r.device(0);
-  if (!r.isNone(0) && device_obj.is_cpu()) {
-    throw std::runtime_error("Invalid device, must be cuda device");
-  }
-  int32_t device_index = -1;
```
Comment (goldsborough) [inline tools/autograd/templates/python_variable_methods.cpp]: If I remember correctly, this method and `dispatch_type_conversion` have some complex logic around `device_index` that we need to verify is preserved here.

File: aten/src/ATen/native/TensorFactories.cpp [hunk 5]
```diff
@@ -595,6 +595,22 @@ Tensor hann_window(
...
+
+Tensor cuda(const Tensor& self, int64_t device_id, bool non_blocking) {
+  if (device_id == -1) { // default argument
+    return self.to(kCUDA, non_blocking);
+  } else {
+    return self.to({kCUDA, device_id}, non_blocking);
+  }
+}
+
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (goldsborough) [inline aten/src/ATen/native/TensorFactories.cpp]: I don't know why I didn't put `~~~ tensor ~~~` above here before, but please write `tensor` where there is currently a space

File: aten/src/ATen/native/TensorFactories.cpp [hunk 6]
```diff
@@ -595,6 +595,22 @@ Tensor hann_window(
...
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ conversions ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Tensor cpu(const Tensor& self) {
+  return self.to(kCPU);
+}
+
+Tensor cuda(const Tensor& self, int64_t device_id, bool non_blocking) {
+  if (device_id == -1) { // default argument
```
Comment (goldsborough) [inline aten/src/ATen/native/TensorFactories.cpp]: This is not necessary, `Device` accepts `-1` as the default index (current device). This line `return self.to(kCUDA, non_blocking);` is actually equivalent to `return self.to({kCUDA, -1}, non_blocking);`, so the whole function should just be `return self.to({kCUDA, device_id}, non_blocking);`.

File: aten/src/ATen/native/TensorFactories.cpp [hunk 7]
```diff
@@ -595,6 +595,22 @@ Tensor hann_window(
       window_length, periodic, /*alpha=*/0.5, /*beta=*/0.5, options);
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ conversions ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Tensor cpu(const Tensor& self) {
+  return self.to(kCPU);
+}
+
+Tensor cuda(const Tensor& self, int64_t device_id, bool non_blocking) {
```
Comment (goldsborough) [inline aten/src/ATen/native/TensorFactories.cpp]: I called it `device_index` everywhere, let's stick to that naming

File: tools/autograd/templates/python_variable_methods.cpp (hunk 4 above)
Comment (goldsborough) [reply tools/autograd/templates/python_variable_methods.cpp]: Yeah so if you follow the logic flow here, the semantics of this method are (cf https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/tensor_conversion_dispatch.cpp#L24):
1. If the device passed to `cuda(...)` supplies an index, the tensor will be allocated on that device,
2. Else, allocate on the device index of `self_`.

You are changing the semantics to:
1. If the device passed to `cuda(...)` supplies an index, the tensor will be allocated on that device,
2. Else, allocate on the current device.

Those two are not equivalent. A quick workaround would be to change the native function to (conceptually): 
```
return self.to({at::kCUDA, device_index == -1 ? (self.is_cuda() ? self.get_device() : -1) : device_index}, non_blocking);
```
but now what if the user actually did pass `Device(at::kCUDA, -1)`, and thus wanted the new tensor to be allocated on the default device? Then with the above scheme we'd be breaking it. That's why `dispatch_type_conversion` used an `at::optional<int32_t>` and then did funny logic around it.

**Review by goldsborough — COMMENTED** [review]
Shouldn't we wait to change the Python binding until the next release? Maybe keep `THPVariable_cuda`, but call `self_.to(...)` inside it, after the `PythonArgParser` took care of handling the deprecated `async` kwarg.

Also, this PR actually changes the semantics of the device on which the new tensor is allocated. See my comment below. I feel like I worked on this method when I did the big `TensorOptions` change and found that the nasty logic inside `dispatch_type_conversion` made things hard.

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (goldsborough) [reply aten/src/ATen/native/native_functions.yaml]: Actually I don't know if you wanted the `optional` here to represent the default device as `nullopt`, or to support the logic of `dispatch_type_conversion`. We would need it for the latter.

File: tools/autograd/templates/python_variable_methods.cpp (hunk 4 above)
Comment (ezyang) [reply tools/autograd/templates/python_variable_methods.cpp]: Yeah, this semantics problem kills this patch. So I guess we have to do `at::optional` here, which means we probably will have to solve `at::optional` before we can fix it. Blah.

**Comment by ezyang** [comment]
Abandoning this patch, it's blocked on `at::optional`ifying the API.

**Comment by gchanan** [comment]
@goldsborough should we move dispatch_type_conversion into ATen?

**Comment by goldsborough** [comment]
@gchanan yes I think more of such logic should move into ATen.

**Pull request closed without merging**